### ISSUE TITLE
LIN_ADVANCE freeze

### ISSUE THREAD

**Issue by nzinov** [MarlinFirmware/Marlin#5699]
I am running d3cb1a86 on delta. When I try to print with ```LIN_ADVANCE``` uncommented, printer moves to the start point, then extruder produces some noise for ~1sec and then printing completely stops. Hotend led is blinking at changing rate as usual - so firmware is not crashed at that moment. However, it only outputs ```busy: processing``` into console and is not responding to any commands. Any tips on debugging that?

**Comment by psavva** [comment]
Can you lower your print speed, say to 35mm/s…?

**Comment by mottihoresh** [comment]
I think the print/move speed may cause it, I am experiencing the same issue. Not sure what's going on. It's not happening with RC8BugFix branch

**Comment by Sebastianv650** [comment]
@mottihoresh there is no branch "RC8BugFix" - do you mean the code base tagged with "1.1.0-RC8"?

@nzinov can you post the gcode section +- some lines around the start point where it freezes? A test with a low speed as the noted 35mm/s can't hurt also, maybe it helps to find the reason for it.
You might also try RCBugFix at the state [without my latest LIN_ADVANCE change.](https://github.com/MarlinFirmware/Marlin/tree/83021d82c60f72f9f5abafb7ad60f54296f81ecc)

If somebody knows other issues mentioning such a behaviour, please give me a hint on them.

**Comment by mottihoresh** [comment]
@Sebastianv650 Meant to write RCBugFix (https://github.com/MarlinFirmware/Marlin/tree/RCBugFix)

I actually spend couple of hours yesterday reading through the issue log. my best guest of what is going on is that the ATMega MCU just chokes since it's trying to process data too fast.

I got non print moves at 160mm/s, and 32 steps per minute, using the new bilinear bed leveling, print speed are at 60 or 80mm/s. When reducing the global feedrate to 50% i noticed no issues, however as soon as I switched it back to 100% the MCU entered panic mode and restarted.

I think this issue may be related:
https://github.com/MarlinFirmware/Marlin/issues/4931

My configuration file: https://gist.github.com/mottihoresh/5f51c1d6c07d75c918992125b4e952ea

**Comment by nzinov** [comment]
@Sebastianv650 here is the begining of my g-code
```
M104 S235 ; set temperature
G28
G29
M420 S1 Z10
M109 S235 ; wait for temperature to be reached
G21 ; set units to millimeters
G90 ; use absolute coordinates
M82 ; use absolute distances for extrusion
G92 E0
M106 S178.5
;layer 0.4 0
G1 Z0.400 F12000.000
G1 X36.622 Y6.495 F12000.000
G1 X36.622 Y18.274 E0.72007 F1440.000
G1 X35.880 Y21.412 E0.91717
G1 X33.222 Y26.716 E1.27982
```
I am going to test lowering print speed a bit later. At this moment my issue is not quite like @mottihoresh. He says the MCU is restarted while in my case it apparently is doing a good job of maintaining temperature even after freeze.

**Comment by Sebastianv650** [comment]
I'm also thinking we have two independent problems here. So your printer is freezing after this line:
`G1 X36.622 Y6.495 F12000.000`

Then I don't think it's the print speed, F1440 is 24mm/s which is already really low.

I have two problems: No knowledge about Marlin code related to bed leveling (as far as I know, your G29 means you are using it) and no knowledge about delta code handling.

What happens if you try a print without bed leveling, is this possible?

If yes, it would be interesting to know if it works without bed leveling and if yes, does it also work with `LIN_ADVANCE` enabled. This way it should be possible to get the problematic feature.

**Comment by nzinov** [comment]
@Sebastianv650 I've tried printing with bed-leveling turned off. Now it makes two circles of skirt with extruder producing strange noise (can't tell for sure but maybe step skip) and then stops in the same manner. Freezes outputting "busy: processing" not responding to commands from host (unfortunately I don't own an LCD)

**Comment by psavva** [comment]
@nzinov If you turn off LIN_ADVANCE, I guess you don't have the issue? right?
Can you provide you config files in a pastebin, for both Configuration.h, and Configuration_adv.h?
Lastly, If you're using 1/32 step drivers, I suggest to configure your control board to 1/8 or max 1/16 steps, and try the test again.

If lowering the step resolution resolves this issue, then I think it certainly has to do with draining the CPU, and thus failure at some point....

**Comment by nzinov** [comment]
@psavva Yes, without LIN_ADVANCE I have no issues (multi-hour prints went OK). I use Polulu drivers, so 1/16 steps max. Here lies my configuration: https://gist.github.com/nzinov/ea6544ce3df78a9438c81bf9b1a435a3.

Besides, I really doubt that CPU would fail to calculate moves but still happily chat to console and handle temperature

**Comment by Sebastianv650** [comment]
I will try to provide you a modified stepper or planner file in the next 1-2 days that outputs some debug data for each move. That should help to find the error.

**Comment by nzinov** [comment]
@Sebastianv650 OK, looking forward to it.)

**Comment by hexane360** [comment]
I'm having a very similar problem. Sending `M905 K0` fixes the problem. I am also using ABL (linear).

I don't have a LCD, so I'm not quite sure what is happening with the firmware. Help debugging would be appreciated.

Additionally, my extruder motor gets very hot during the resulting fugue state, and sending M112 doesn't seem to have an effect.

**Comment by hexane360** [comment]
I've been playing around with various things, and the problem seems to be *VERY* weird.

Some general printer information:
-Lulzbot Mini
-Cartesian
-ABL Linear
-1/8 Microstepping (down from 1/16)
-Latest RCBugFix

And here is the start G-Code that caused the problem:
```
G21                          ; metric values
G90                          ; absolute positioning
M82                          ; set extruder to absolute mode
;M200 D2.9175 T0 ; set filament diameter
M200 D0 T0; cancel volumetric extrusion
M107                         ; start with the fan off
G92 E0                       ; set extruder position to 0
M140 S55                     ; get bed heating up
G28                               ; home all
M109 S150                     ; set to cleaning temp and wait
G1 Z150 E-10 F75           ; suck up XXmm of filament
M109 S170                     ; heat up rest of way
G12  P1 S2                      ; wipe
G01 Z20 F700              ; raise off wipe pad
M109 S170                    ; set to probing temp
M204 S300                    ; set accel for probing
G29                          ; Probe
M204 S2000                   ; set accel back to normal
G1 X5 Y15 Z10 F5000          ; get out the way
M400                         ; clear buffer
G4 S1                        ; pause
M109 S190    ; set extruder temp and wait
M190 S55 ; set bed temp and wait
G1 Z2 E0 F75                 ; extrude filament back into nozzle
G21 ; set units to millimeters
G90 ; use absolute coordinates
M82 ; use absolute distances for extrusion
G92 E0
;layer 0, z:0.4
G1 E-1.00000 F600.00000
G92 E0
G1 Z0.200 F10800.000
G1 Z0.400 F10800.000
G1 Z0.600 F10800.000
G1 X64.958 Y65.166 F10800.000
G1 Z0.400 F10800.000
G1 E1.00000 F600.00000
M204 S500
G1 F1800
```
No amount of manually entering commands allowed me to recreate the problem. However, if I started the print with `M905 K0` and then paused, entered `M905 K25`, and resumed, the print had no trouble. Eventually, I started stripping out extruder moves until I got it working. One of the following lines caused it to fail:
```
G1 Z2 E0 F75
G92 E0
G1 E-1 F600
G1 E1 F600
```
It seems like the culprit could be one of the following:
- Slow feedrates
- Sign of extrusion (- to + or + to -)
- Something else

Disabling retraction in Slic3r also removes the problem code. However, I don't think that's the root cause.

**Comment by hexane360** [comment]
@Sebastianv650 I've been poking through planner.cpp getting an idea how LIN_ADVANCE works. The method Planner::_set_position_mm seems to reset the position[] array. However, position_float is only updated after a move has taken place, using `memcpy(position_float, target_float, sizeof(position_float));`. Is it possible that the first planner move after _set_position_mm is called will use a stale position_float[]?
```
void Planner::_set_position_mm(const float &a, const float &b, const float &c, const float &e) {
  #if ENABLED(DISTINCT_E_FACTORS)
    #define _EINDEX (E_AXIS + active_extruder)
    last_extruder = active_extruder;
  #else
    #define _EINDEX E_AXIS
  #endif
  long na = position[X_AXIS] = lround(a * axis_steps_per_mm[X_AXIS]),
       nb = position[Y_AXIS] = lround(b * axis_steps_per_mm[Y_AXIS]),
       nc = position[Z_AXIS] = lround(c * axis_steps_per_mm[Z_AXIS]),
       ne = position[E_AXIS] = lround(e * axis_steps_per_mm[_EINDEX]);
  stepper.set_position(na, nb, nc, ne);
  previous_nominal_speed = 0.0; // Resets planner junction speeds. Assumes start from rest.
  ZERO(previous_speed);
}
```
There may also be other gaps similar to this. Looking through the file, there's a lot of places where `position[]` is set without `position_float[]`.

**Comment by Sebastianv650** [comment]
I don't think that's a problem, but I can't say that for 100% because I don't know where all that set_position things are called. According to some comments, they are used for homing, probing and such things. That would mean they will not be active during the main print.
But all freezes I know of happen during a print, not during homing?

I'm still having no idea where this problem comes from. Even with a gcode from another TAZ 5 that produces this issue, I can't reproduce it on my TAZ 5. It seems to be a very specific problem that happens only with specific configs combined with specific numbers in the gcode.
At the moment my only hope is that some day somebody can provide a gcode snipplet that lets me reproduce the error.. :-(

**Comment by billyd60** [comment]
Sebastian, I have solved most of my issues with RCBUGFIX and it was mostly hardware related. I had a bad enclosure connector on my heated bed cable and the silicone heater was going bad as well. Crazy to have two problems so closely related. Anyway I have a new silicone heater and I am running the cables (power and thermistor for the bed) direct into the RAMBO with no breaks. This is actually amazingly good as I think I was getting some serious losses in power and accuracy with the two standard breaks (Enclosure connector and at the bed) that the Taz 5 has in this line. My bed has never heated so fast or been this stable. So I may have had a balky enclosure connector from day1

Anyway long story short I still get a freeze in printing from time to time (pretty rare though). Other than that lin advance and Rcbugfix are working great and producing the best prints I've ever gotten. The freeze is bizarre though, the printer just sits there, heat still going to the extruder and bed, lcd screen reading normally, just zero movement. Just happened again last night for the first time in weeks, printing in PLA. It seems to happen on large complex models. Maybe just because they take so long. Never have any trouble with small prints.

**Comment by Sebastianv650** [comment]
Glad to hear that at least some problems were hardware dependent!

When it freezes, is the LCD menu also functional, meaning it's possible to enter and navigate through it? If yes I will continue searching around the stepper ISR handler.

**Comment by billyd60** [comment]
I had to cycle power to get the extruder off the part (to use the prepare menu move z axis) so the LCD screen was not responsive to the dial control. The extruder and heat bed were still at temperature though, and the LCD screen looked normal in every respect and it was still reading the thermistor temps as well as x y and z coordinates which lined up with where the extruder was approximately (I didn't actually measure just by eye).

**Comment by Sebastianv650** [comment]
I still have no real answer, but you might try the following with a code that usualy crashes:

Inside void Stepper::ISRhandler(), comment out the lines containing sei() and cli().
Inside void Temperature::isr(), comment out the line containing sei().

This will undo some ISR handling changes I did to Marlin. You might see a lot of serial communication errors this way, so you might print at a lower serial baud rate for this test.

Second option: Undo the changes from above, but add a cli(); just above SBI(TIMSK0, OCIE0B); at the end of void Temperature::isr().

Both changes are "a shot in the dark" if my translator was able to translate what I mean ;-) But at least it's better than nothing..

**Comment by billyd60** [comment]
I only print from SD card never from USB.

**Comment by billyd60** [comment]
Sebastian I believe I have isolated the freeze problem. If I stop a print prematurely from the LCD panel and restart the job without cycling power to the RAMBO, the printer will ultimately freeze mid print at some point. It also cause odd sudden drops in the extruder temp and bed temp readings. None of this behavior occurs if I simply start the printer and print a job. Stopping the print from the LCD panel seems to cause a whole host of problems with RCBUGFIX.

All of this is just what I have observed in using the firmware, I can't prove any of it since I can't code.

I hope this puts you on the right path to find out why lin_adv freezes a print job sometimes.

**Comment by Sebastianv650** [comment]
Thanks, i will try to reproduce that!

**Comment by Sebastianv650** [comment]
I started a print from SD card, let it run for a minute or so and then aborted it.
Restarted it, let it run for 1-2 minutes. No crash.

How long should it take until the freeze? As there is nothing "counting up" within the stepper or advance code, I have no idea why it should behave different when I let it run longer..
I had a quick look at the issue page (I wasn't activly reading it for some weeks) and there seems to be quite some freezing, temp. jumps and other issues with the recent code page. That make me believe the error might not be in LIN_ADVANCE specific because they are not mentioning it. Maybe advance only increases the probability the error shows up?

**Comment by thinkyhead** [comment]
I think there's probably some merit to the point that `position_float` and other additions for `LIN_ADVANCE` are not being set properly. Note that `position` itself is never updated until the end of `_buffer_line`. However the `LIN_ADVANCE` code updates it right away, at the top of `_buffer_line`. Likewise the E movement is skipped when in dry-run mode, but this is after the `LIN_ADVANCE` code is using it. So I think the `LIN_ADVANCE` code in `_buffer_line` needs some tweaks to make sure `position_float[]` has full parity with `position[]`.

**Comment by Sebastianv650** [comment]
I thought I only doubled the position array by the position_float thing. It looks like I have to dig into the code again and compare what both of them are doing once more.

**Comment by Sebastianv650** [comment]
@billyd60 Can you try this branch and see if it solves for example your print from SD -> abort & restart problem? It's RCBugFix + a commit that should cleanup all position <> position_float things.
https://github.com/Sebastianv650/Marlin/tree/cleanup-position_float

If it works better now, I would create a PR.

**Comment by billyd60** [comment]
Hi Sebastian been away from my computer for awhile. Yes I will try it tonight or tomorrow.

Regarding the sudden drops in the measured thermistor temps (after an lcd panel stop print command and restart) it appears as though the code is "losing track" of the actual readings from the thermistors (Bed and extruder) periodically. Ultimately I believe the temperature drops so low that it may trigger cold extrusion protection. This is just a guess though.

Btw this behavior sometimes takes a few hours to show up (both the odd temperature drops and print freeze). So it is quite difficult to reproduce. This suggests a very slow build up of the problem. Perhaps some kind of additive floating point problem that takes tens of thousands of iterations to show up? Again this behavior never happens on my printer without an LCD stop print command and then restart without cycling power. As long as I turn on the printer, and execute a print job without cancelling it, I have never had a print fail. I just completed a 9.5 hour print job with Rcbugfix and it was flawless.

**Comment by Sebastianv650** [comment]
@billyd60 please read the last posts from issue #5698, this might be another possible reason for the freezes you have.

**Comment by billyd60** [comment]
Do I need to use your config and adv.h files and change them for the Taz 5 or can I just use my current config and _adv.h files I am currently using for Rcbugfix? I really don't feel like doing all that editing again...

**Comment by Sebastianv650** [comment]
My two PR are not changing the config files. Depending on how old your current ones are, you might be able to copy and paste them. I would open your current one and the maybe newer RCBugFix one in notepad++ and use the compare function to see if there are structural changes. If not, just use the old ones.

**Comment by psavva** [comment]
I just simulated the same problem.
Printer stopped mid print, and froze.

I'm using LIN_Advanced as well. K=400.
The LCD screen RRD Full Graphics Smart Controller became extremely slow...
The reset button would not respond.
The LCD menu items didn't have an effect.

I'm trying to print the same thing again, and see if its reproducible...

I'll merge your PR in a local branch and try again to see if its fixed (if I manage to replicate it)

**Comment by psavva** [comment]
I replicated the issue again.
I had K=300, and had no problem.
I set K=400, and the printer froze before it finished a 10cm scirt.

I upped the the FR to 150%, and instantly froze...

**Comment by Sebastianv650** [comment]
With the pr to prevent temperatur isr reentry?

**Comment by psavva** [comment]
Not yet, now I know how to trigger, I'll try again with the PR

**Comment by Sebastianv650** [comment]
Ah, I was afraid it was with the PR :-)
The behaviour is logical: When you fire the stepper ISR more often, which is true with a high K value and/or a high FR, there is a higher chance to trigger a stepper ISR within a temperature ISR. Now let's hope the best it's now gone!

**Comment by psavva** [comment]
@Sebastianv650, The problem still exists, even when applying https://github.com/Sebastianv650/Marlin/tree/cleanup-position_float

I just simulated in a matter of seconds.

I'm using:
DRV-8825's at 1/32 steps.
Lin_Advance; K=400
Bilinear Bed Levelling
50 mm/s Print Speed, but set FR to 300% to simulate the issue.

Update:
I just turned Lin_Advanced off, and printed the exact same gcode, at 600% FR, with no freezing...

**Comment by Sebastianv650** [comment]
You need another PR, this one: https://github.com/Sebastianv650/Marlin/tree/Protect_tempISR
The one you applied can also prevent problems, but only in rare situations I think.

**Comment by psavva** [comment]
Hi @Sebastianv650 
I'm sorry to report that the problem still exists after merging Protect_tempISR.

I went to 250% FR, and Froze immediately.

I turned off LIN_ADVANCE, and no issue at 600% FR.

Let me know if I can do anything else to help diagnose the issue.
I'm going to retire for tonight, early next week, or maybe over the weekend, I will be able to test it again, unless @billyd60 can try test before then :)

**Comment by thinkyhead** [comment]
> the problem still exists after merging Protect_tempISR

If that doesn't address this, then probably #5829 doesn't either. But give it a check. I have a query in about the line `cli();` at the end of `Temperature::isr` — maybe it's connected, but I can't confirm.

**Comment by Sebastianv650** [comment]
@psavva can you give me the speed your print is running (at 100% FR) and your steps/mm setting for x,y and e axis? I want to see what step rate we speek about when it's freezing.
I think your problem might be something different than re-entry of temperature ISR, as your freeze isn't giving you a thermal error message and it doesn't happen with LIN_ADVANCE disabled. But that's also an advantage, this limits the areas where I have to search.

I will be blocked until friday, but I should find some free minutes in between to read through the code and think about it anyway.

**Comment by psavva** [comment]
@Sebastianv650 
Print Speed: 70mm/s @ 100% FR
Configuration.h
http://pastebin.com/XBmAGqzq

Configuration_Adv.h
http://pastebin.com/2D2uvRE1

When it freezes, I do not get a Thermal Error Message.
The Steppers just stop.
The LCD is extremely slow, but still functional (somewhat)

**Comment by Sebastianv650** [comment]
> The LCD is extremely slow, but still functional (somewhat)

That's good to know, and also what I expected. It's clearly something with the stepper ISRs with LIN_ADVANCE at high step rates.

With your settings, your leading axis has 11200steps/second, which already means double stepping to get it done. Just for information, at 600% FR (67.2ksteps/s) you will most likely don't see 6x the speed of 100% FR because Marlin is already just handling the steps as fast as the CPU can do it.
I will have a look what's going wrong with LIN_ADVANCE..

**Comment by Sebastianv650** [comment]
@psavva a quick check for someone that can reproduce the error - I can't, just tested it at 600% FR and no crash :-/

Have a look into stepper.cpp, search for the `void Stepper::ISRhandler()` block.
At the moment, it should say:
```cpp
    if (!nextMainISRrun) {
      isr();
    }
    if (!nextAdvanceISRrun) {
      advance_isr();
    }
    
    if (nextAdvanceISRrun <= nextMainISRrun) {
      OCR1A = nextAdvanceISRrun;
      if (nextMainISRrun)
        nextMainISRrun -= nextAdvanceISRrun;
      nextAdvanceISRrun = 0;
    }
```
change it to:
```cpp
    if (!nextMainISRrun) {
      isr();
    }
    if (!nextAdvanceISRrun) {
      advance_isr();
    }

    //Start Debugging Code
    if (!nextMainISRrun) {
      SERIAL_ECHO_START;
      SERIAL_ECHOLNPGM("nMainISRrun=0");
      nextMainISRrun = 65535;
    }
    if (!nextAdvanceISRrun) {
      SERIAL_ECHO_START;
      SERIAL_ECHOLNPGM("nAdvanceISRrun=0");
      nextAdvanceISRrun = 65535;
    }
    //End Debugging Code
    
    if (nextAdvanceISRrun <= nextMainISRrun) {
      OCR1A = nextAdvanceISRrun;
      if (nextMainISRrun)
        nextMainISRrun -= nextAdvanceISRrun;
      nextAdvanceISRrun = 0;
    }
```

This is not for printing, but if one of the two values become 0 that might be a reason for the behaviour of your printer. If it is, you should get a message over serial when it's crashing.
Please test this when you find the time and report if it gives you one of the messages and which one.

**Comment by psavva** [comment]
I'll report back by tomorrow :)

**Comment by billyd60** [comment]
Sorry I have not been able to do any testing. Too many print jobs to mess with it at the moment. Actually I am going to bow out of this because Rcbugfix is working perfectly for me now, all I have to do is not cancel a print job from the lcd panel or if I do, make sure to power cycle the Rambo before running a print and I get perfect prints every time. So I will remain a spectator and see what you guys come up with to solve it.

**Comment by psavva** [comment]
@Sebastianv650 the source code doesn't exist in either stepper.cpp, or anywhere else in the codebase in RCBugFix to debug as instructed.

Please double check what I should do to debug, and i'll do so.

**Comment by Sebastianv650** [comment]
You are right, the code has changed (renamed) in RCBugFix. Goto stepper.cpp, the code block is now "`void Stepper::advance_isr_scheduler()`".

Current Code:
```cpp
    // Run main stepping ISR if flagged
    if (!nextMainISR) isr();

    // Run Advance stepping ISR if flagged
    if (!nextAdvanceISR) advance_isr();

    // Is the next advance ISR scheduled before the next main ISR?
    if (nextAdvanceISR <= nextMainISR) {
      // Set up the next interrupt
      OCR1A = nextAdvanceISR;
      // New interval for the next main ISR
      if (nextMainISR) nextMainISR -= nextAdvanceISR;
      // Will call Stepper::advance_isr on the next interrupt
      nextAdvanceISR = 0;
    }
```
change to:
```cpp
    // Run main stepping ISR if flagged
    if (!nextMainISR) isr();

    // Run Advance stepping ISR if flagged
    if (!nextAdvanceISR) advance_isr();
    
    //Start Debugging Code
    if (!nextMainISR) {
      SERIAL_ECHO_START;
      SERIAL_ECHOLNPGM("nextMainISR=0");
      nextMainISR= 65535;
    }
    if (!nextAdvanceISR) {
      SERIAL_ECHO_START;
      SERIAL_ECHOLNPGM("nextAdvanceISR=0");
      nextAdvanceISR= 65535;
    }
    //End Debugging Code

    // Is the next advance ISR scheduled before the next main ISR?
    if (nextAdvanceISR <= nextMainISR) {
      // Set up the next interrupt
      OCR1A = nextAdvanceISR;
      // New interval for the next main ISR
      if (nextMainISR) nextMainISR -= nextAdvanceISR;
      // Will call Stepper::advance_isr on the next interrupt
      nextAdvanceISR = 0;
    }
```

**Comment by psavva** [comment]
It's not freezing anymore!!!
http://pastebin.com/y4zJ4ake

I don't get it.
From being able to consistently getting it to freeze going to FR250%, now it doesn't

I don't see what could have changed.

It also froze on me without LIN_ADVANCE enabled this morning.

Only difference now, is that the serial monitor is running, whereas before, I did not have a laptop connected, printed directly from the SD Card, without the Serial Monitor.

**Comment by psavva** [comment]
The printer just froze again, but this time with a Thermal Runaway on the LCD.  No laptop connected, and LIN_ADVANCE enabled

**Comment by Sebastianv650** [comment]
I just recognised you are running a dual extruder setup. But on a first look, the code looks OK with that.
The reaction is as I expected it. Your printer is not freezing because this line
```cpp
nextAdvanceISR= 65535;
```
is disabling the advance ISR before it will crash your printer. If `nextAdvanceISR` becomes 0, this code block will "freeze" the time for the step execution because it subtracts 0 from `nextMainISR `.
```cpp
    // Is the next advance ISR scheduled before the next main ISR?
    if (nextAdvanceISR <= nextMainISR) {
      // Set up the next interrupt
      OCR1A = nextAdvanceISR;
      // New interval for the next main ISR
      if (nextMainISR) nextMainISR -= nextAdvanceISR;
      // Will call Stepper::advance_isr on the next interrupt
      nextAdvanceISR = 0;
    }
```
So the good news is, we now know why it freezes printing and we also know why the display is still function, but slow: The printer is flooded with extruder ISRs at the max. possible rate.
The bad news is, `nextAdvanceISR` shouldn't be able to become 0. When I did the coding for that, I had such a debuging code running on my printer and never got a message.
The extruder stepper rate is calculated with
```cpp
#define ADV_RATE(T, L) (e_steps[TOOL_E_INDEX] ? (T) * (L) / abs(e_steps[TOOL_E_INDEX]) : ADV_NEVER)
```
so there can't be a division by 0. So there are only 3 ways to get a 0 result:
- The timer value T is 0. Can't be, it's limited to reasonable values.
- The steps/loop count L is 0. Can't be, it can only be 1, 2 or 4.
- e_steps value is very high, so the rounded result becomes 0. This should not happen, but you are using a much higher K value than I ever tested (400), so it might be possible at high speeds.

Let's modify the debug code to see what values we have when we get a 0. Copy the following code in a new line after every
```cpp
eISR_Rate = ADV_RATE(
```
in the stepper.cpp:

```cpp
    //Start Debugging Code
    if (!eISR_Rate) {
      SERIAL_ECHO_START;
      SERIAL_ECHOPAIR("Timer:", timer);
      SERIAL_ECHOPAIR("step_loops:", step_loops);
      SERIAL_ECHOPAIR("esteps:", e_steps[TOOL_E_INDEX]);
      SERIAL_ECHOPAIR("tool_index:", TOOL_E_INDEX);
      eISR_Rate= 65535; //Disable the extruder ISR to prevent a frozen system.
    }
    //End Debugging Code
```
When you get a handfull of lines, paste it as you have done before and we will see..

> The printer just froze again, but this time with a Thermal Runaway on the LCD. No laptop connected, and LIN_ADVANCE enabled

This time also with the debug code or without?

**Comment by psavva** [comment]
It was with the debug code...

**Comment by Sebastianv650** [comment]
> It was with the debug code...

I guess it's due to the same error. We checked for 0 with the debug code, but if it becomes another tiny number it can also block the system that much that the heater system doesn't have enough calculation time to do it's job.

Thanks for your help so far, we should find a solution.

**Comment by psavva** [comment]
@Sebastianv650 I think I need to make a small change to the debug code in a specific case:

After the specific line:
 eISR_Rate = ADV_RATE(OCR1A_nominal, step_loops_nominal);

Timer is not recognised in the scope in the debug clause
I'll change that to:

   #if ENABLED(LIN_ADVANCE)

      if (current_block->use_advance_lead)
        current_estep_rate[TOOL_E_INDEX] = final_estep_rate;

      eISR_Rate = ADV_RATE(OCR1A_nominal, step_loops_nominal);
	 //Start Debugging Code
		if (!eISR_Rate) {
		  SERIAL_ECHO_START;
		  SERIAL_ECHOPAIR("OCR1A_nominal:", OCR1A_nominal);
		  SERIAL_ECHOPAIR("step_loops:", step_loops);
		  SERIAL_ECHOPAIR("esteps:", e_steps[TOOL_E_INDEX]);
		  SERIAL_ECHOPAIR("tool_index:", TOOL_E_INDEX);
		  eISR_Rate= 65535; //Disable the extruder ISR to prevent a frozen system.
		}
		//End Debugging Code
    #endif

    SPLIT(OCR1A_nominal);  // split step into multiple ISRs if larger than  ENDSTOP_NOMINAL_OCR_VAL
    _NEXT_ISR(ocr_val);

Since the timer for ADV_RATE is OCR1A_nominal in this case...

Ok?

**Comment by Sebastianv650** [comment]
Yes, that's right!

**Comment by psavva** [comment]
@Sebastianv650 

Here's the results:

I have deleted the "echo:busy: processing" in the log file...
Find the paste here :)
http://pastebin.com/906NKWUS

**Comment by Sebastianv650** [comment]
Alright, we should not wonder when Marlin crashes when it tries to execute over 4 million steps per second. That's what your log file says ;-) The maximum Marlin can do on an ATMega is 40.000 steps per second.

That high extruder step amount should be a result of the high K factor. I would double check if it's realy needed, but nevertheless we need a solution to handle such cases without crashes.
Let's think about what options we have:
- Write a check function that tries to find out if a K value is save after setting it. If it's not, throw a message and prevent printing. If you don't have a serial connection, the printer will do nothing and you don't have an idea why. And I think there is no perfect way of determining if K is valid or not. Bad.
- Determine a safe maximum extruder step rate (OCR1A value calculated from maximum extruder speed setting). If the calculated eISR step rate is lower, set it to the min. allowed OCR1A value. This will delay the extruder advance steps a little bit, but no one should get lost and the printer keeps alive. Beside that, a boolean variable could be set if this safety feature was used, printing a warning message compareable to the Busy printing.. message from time to time so the user knows (if connected..) that the advance steps are not longer executed 100% where they should go in time.

I will try to reproduce this again using a high K value combined with high speed to be sure it's no other calculation in the LIN_ADVANCE process that produces wrong numbers before starting with a patch for this.

**Comment by psavva** [comment]
Let me know if there is anything else you need from me, happy to oblige :)

Seriously awesome work...
Now when 32-Bit Marlin is up, maybe we can get 4 million steps per second :)

**Comment by Sebastianv650** [comment]
I applied your steps/mm settings and also jerk and acceleration values. No freezing, no 0 values (tested up to 180mm/s). I have no clue why I can't get your result in my printer..
I'm afraid that the high esteps values are maybe the result of an overflow or something and that limiting the eISR rate will cure only the result but not the root cause.

I think I have to use your printer again. I will create a branch of Marlin and modify the code to track all values used for LIN_ADVANCE. When it reaches the point where we have that magic 0, it will dump all the data to serial and block all further movements until the printer is power cycled to prevent a flooded serial display.
That will need some time for coding, but with this final test I should be able to follow the complete calculation. Therefore if something overflows I will see it, and if not it should be clear why your code ends up with no time between ISRs while mine never goes below 21 clock cycles.

> Now when 32-Bit Marlin is up, maybe we can get 4 million steps per second :)

If you find a stepper that can handle that speed :D With your estep/mm setting of 385, that means over 10m per second.

**Comment by Sebastianv650** [comment]
I need your patience again. Here is the code with the debug information: https://github.com/Sebastianv650/Marlin/tree/Debug_esteps

Download it, copy and paste your configuration.h and configuration_adv.h files into it and give it a try. The code compiles on my machine, but everything else is untested, so fingers crossed that it throughs out a lot of numbers.
And fingers crossed that we don't need another 10 loops to solve this! :)

**Comment by psavva** [comment]
@Sebastianv650 , Need a little help to compile.

I guess int* is a pointer?
I'm not a C++ Programmer, and thus don't know how to fix it ...

Arduino: 1.6.8 (Windows 8.1), Board: "Arduino/Genuino Mega or Mega 2560, ATmega2560 (Mega 2560)"

sketch\stepper.cpp: In static member function 'static void Stepper::isr()':
stepper.cpp:619: error: invalid conversion from 'int*' to 'int' [-fpermissive]
       debug_current_estep_rate = current_estep_rate;
                                ^
stepper.cpp:620: error: invalid conversion from 'int*' to 'int' [-fpermissive]
       debug_current_adv_steps = current_adv_steps;
                               ^
exit status 1
invalid conversion from 'int*' to 'int' [-fpermissive]
This report would have more information with
"Show verbose output during compilation"
option enabled in File -> Preferences.

**Comment by psavva** [comment]
I guess it should be
debug_current_adv_steps =
current_estep_rate[TOOL_E_INDEX];
debug_current_adv_steps = current_adv_steps[TOOL_E_INDEX];

Can you confirm before I run the next test?

**Comment by Sebastianv650** [comment]
Yes, that was wrong. I updated it in the repository: https://github.com/Sebastianv650/Marlin/tree/Debug_esteps

**Comment by psavva** [comment]
And here you go.

Crashed during crusing...
http://pastebin.com/rNEMAWF3

**Comment by psavva** [comment]
I'm busy running a second test, where i'm printing this model: http://www.thingiverse.com/thing:1624412/#files

Lots of small moves.
I'm checking the serial monitor, and had a "cold extrusion prevented", and i'm 9 mins into the print.
I'll upload the log file once it's finished or frozen ;)

**Comment by psavva** [comment]
And Bingo..  Crashed...

http://pastebin.com/ccB2MkEU

Hope this is enough to find the issue :)
This time is crashed during decel...

**Comment by Sebastianv650** [comment]
Thanks again!
You are using 1.75mm filament with a layer height around 0.2-0.25mm in this prints? Just trying to see if the calculations regarding the extruder speed <=> movement speed connection can be right.

Both crash cases are not logical to happen on a error free printer:
- During crusing, even on the first step, there should be no speed change to the last step. Therefore no advance steps can happen. Comming from 75 advance steps executed, your printer tries to add another 549 where I would expect 0.
- During deceleration, the extruder speed can only slow down. Therefore the amount of executed advance steps has to be reduced. But your printer, comming from 294 executed advance steps, wants to increase the nozzle pressure even more in this case by adding another 234 steps instead of subtracting some steps.

The calculation itself is OK, nothing seems to overflow.
There is basicaly only one thing left which would also explain why I can't reproduce your issue: While I used all you settings like steps/mm, acceleration and so on, I was not using bed leveling as you do. I was never using it, therefore I also don't know how to set it up for a quick test.
Therefore I also don't know exactly what it does to the moves inside the planner.

Guess what comes next?
I'm sorry to keep asking such things, but would it be possible for you to disable the bed leveling thing in Configuration.h completely and re run the known for crashing code again, maybe including a z-offset to be shure your nozzle isn't damaging the bed?
If it's not crashing now, we have a winner. Then I have to dig into the bits and bytes of bilinear leveling, set it up for my printer and I should also have a crashing system. That's my goal, I don't see another way to debug this only based on remote tests - and it's also the right way not to eat other peoples time..

**Comment by psavva** [comment]
@Sebastianv650 , it is always my pressure to work with this community :)
I will certainly test and report back to you soon...

**Comment by psavva** [comment]
Hi @Sebastianv650, 
It failed with LIN_ADVANCE enabled, and all Bed Levelling Features disabled.

http://pastebin.com/HN1CHiPk

I'm about to test with LIN_ADVANCE Disabled, and Bed Levelling Disabled, and report back shortly

**Comment by psavva** [comment]
Hi @Sebastianv650, 
With `LIN_ADVANCE` disabled, and FR to 900%, then also cranked up Velocity + Acceleration via LCD Menu during the print to try push to the MAX, and there is No Failure after 10 Minutes of Printing...
```
start
echo: External Reset
Marlin 1.1.0-RCBugFix

echo: Last Updated: 2016-12-06 12:00 | Author: (none, default config)
Compiled: Feb 23 2017
echo: Free Memory: 3238  PlannerBufferBytes: 1232
echo:V29 stored set  M203 etrieved (393 bytes)
echo:Steps per unit:
echo:  M92 X160.00 Y160.00 Z800.00 E385.00
e00 Z500 E3000
echo:Accelerati
echo:  M203 X250.00 Y250.00 Z20.00avel
echo:  M204 Pum Acceleration (mm/s2):
echo:  M201 X2000 Y2000 Z500 E3000
echo:Accelerations: P=printing, R=retract and T=travel
echo:  M204 P2000.00 R3000.00 T750.00
echo:Advanced variables: S=Min feedrate (mm/s), T=Min travel feedrate (mm/s), B=minimum segment time (ms), X=maximum XY jerk (mm/s),  Z=maximum Z jerk (mm/s),  E=maximum E jerk (mm/s)
echo:  M205 S0.00 T0.00 B20000 X10.00 Y10.00 Z0.40 E10.00
echo:Home offset (mm)
e0 H180 B70 F0
  M145 S1 H240 Bcho:Hotend offsets (mm)
echo:  M21301 P51.83 I5.020
echo:Material heatup parameters:
echo:  M145 S0 H180 B70 F0
  M145 S1 H240 B110 F0
echo:PID settings:
echo:  M301 P51.83 I5.02 D133.75
echo:Filament settings: Disabled
echo:  M200 D3.00
echo:  M200 T1 D3.00
echo:  M200 D0
echo:Z-Probe Offset (mm):
echo:  M851 Z2.00
echo:SD card ok
echo:SD card ok
echo:enqueueing "M23 wb_dra~1.gco"
echo:enqueueing "M24"
echo:Now fresh file: wb_dra~1.gco
File opened: wb_dra~1.gco Size: 4023557
File selected
 T:189.5 /190.0 T0:189.5 /190.0 T1:0.0 /0.0 @:72 @0:72 @1:0 W:8
 T:189.6 /190.0 T0:189.6 /190.0 T1:0.0 /0.0 @:76 @0:76 @1:0 W:7
 T:190.0 /190.0 T0:190.0 /190.0 T1:0.0 /0.0 @:67 @0:67 @1:0 W:6
 T:190.0 /190.0 T0:190.0 /190.0 T1:0.0 /0.0 @:72 @0:72 @1:0 W:5
 T:190.4 /190.0 T0:190.4 /190.0 T1:0.0 /0.0 @:59 @0:59 @1:0 W:4
 T:190.5 /190.0 T0:190.5 /190.0 T1:0.0 /0.0 @:59 @0:59 @1:0 W:3
 T:190.5 /190.0 T0:190.5 /190.0 T1:0.0 /0.0 @:61 @0:61 @1:0 W:2
 T:190.5 /190.0 T0:190.5 /190.0 T1:0.0 /0.0 @:62 @0:62 @1:0 W:1
 T:190.5 /190.0 T0:190.5 /190.0 T1:0.0 /0.0 @:63 @0:63 @1:0 W:0
echo:Active Extruder: 0
echo:busy: processing
echo:busy: processing
echo:busy: processing
echo:Active Extruder: 0
X:160.00 Y:105.00 Z:10.00 E:0.00 Count X:25600 Y:16800 Z:8000
X:160.00 Y:105.00 Z:15.00 E:0.00 Count X:25600 Y:16800 Z:8005
X:160.00 Y:105.00 Z:15.00 E:0.00 Count X:25600 Y:16800 Z:8041
echo:Settings Stored (393 bytes)
echo:Settings Stored (393 bytes)
echo:enqueueing "M84 X Y Z E"
```

**Comment by psavva** [comment]
After 10 Mins of Printing with Bilinear Bed Levelling Enabled, and No `LIN_ADVANCE`, No crashes.
It's certainly caused by `LIN_ADVANCE`, Please let me know if there is anything I can do further to help?

Remote Desktop :) ?

**Comment by Sebastianv650** [comment]
> Remote Desktop :) ?

We need a remote printer connection in Marlin for debugging - great idea ;)

Next thing I would do is to have a look at the gcode you use to reproduce the crash. If you know the area where it crashes it would be even better, but if not I will do the analysis on the complete file. So please upload it :)

If I can't find anything in it, I will set as much as possible in my config to your settings and run exactly that gcode at a high FR. Then it simply has to crash..! Hopefuly..

I still have no real idea where to start for specific tests. We know now what numbers the error produces and that this will freeze the system. But where is this ugly small bug sitting that causes it? A Marlin simulator would be nice, where we could run a specific configuration with gcode and watching some variables step by step..

**Comment by psavva** [comment]
@Sebastianv650 , You can find the gCode here:
https://drive.google.com/open?id=0B6kS8tqbO6jVR2s3ajRRT3FBVGc

I have no idea on what line it fails, as it's really random it seems...

I will now test LIN_AVANCE, with a single extruder.
Maybe Dual Extrusion Setup was causing it. I'll report back later.

**Comment by psavva** [comment]
Hi @Sebastianv650 , I can now confirm that with Dual Extrusion Disabled, it will still fail with `LIN_ADVANCE` enabled.  It's something else...
```
start
echo: External Reset
Marlin 1.1.0-RCBugFix

echo: Last Updated: 2016-12-06 12:00 | Author: (none, default config)
Compiled: Feb 24 2017
echo: Free Memory: 2848  PlannerBufferBytes: 1568
Error:EEPROM checksum mismatch
echo:Hardcoded Default Settings Loaded
echo:Steps per unit:
echo:  M92 X160.00 Y160.00 Z800.00 E385.00
echo:Maximum feedrates (mm/s):
echo:  M203 X150.00 Y150.00 Z6.00 E25.00
echo:Maximum00 R5000.00 n (mm/s2):
echo:  M201 X750 Y750 Z100 E5000
echo:Accelerations: P=printing, R=retract and T=travel
echo:  M204 P750.00 R5000.00 T750.00
echo:Advanced variables: S=Min feedrate (mm/s), T=Min travel feedrate (mm/s), B=minimum segment time (ms), X=maximum XY jerk (mm/s),  Z=maximum Z jerk (mm/s),  E=maximum E jerk (mm/s)
echo:  M205 S0.00 T0.00 B20000 X8.00 Y8.00 Z0.40 E5.00
echo:Home offset (mm)
echo:  M206 X0.00 Y0.00 Z0.00
Auto Bed Leveling:
echo:  M420 S0
echo:Material heatup partings: Disabled
ech5 S0 H190 B50 F0
  M145 S1 H240 B110 F0
echo:fset (ttings:
echo:  M301 P51.83 I5.02 D133.75
echo:Filament settings: Disabled
echo:  M200 D1.75
echo:  M200 D0
echo:Z-Probe Offset (mm):
echo:  M851 Z-0.60
echo:SD card ok
echo:SD card ok
echo:enqueueing "M23 wb_dra~1.gco"
echo:enqueueing "M24"
echo:Now fresh file: wb_dra~1.gco
File opened: wb_dra~1.gco Size: 4023557
File selected

 T:190.5 /190.0 @:59 W:3
 T:190.5 /190.0 @:61 W:2
 T:190.5 /190.0 @:63 W:1
 T:190.5 /190.0 @:63 W:0
echo:busy: processing
X:160.00 Y:105.00 Z:10.60 E:0.00 Count X:25600 Y:16800 Z:8480
G29 Auto Bed Leveling
echo:busy: processing
echo:busy: processing
Bilinear Leveling Grid:
      0     1     2
 0 +0.40 +0.52 +0.51
 1 +0.27 -0.24 -0.80
 2 +0.70 -0.48 -1.60

X:340.00 Y:290.00 Z:11.81 E:0.00 Count X:54400 Y:46400 Z:8480
X:340.00 Y:290.00 Z:15.00 E:0.00 Count X:54400 Y:46400 Z:8482
X:340.00 Y:290.00 Z:15.00 E:0.00 Count X:54400 Y:46400 Z:8490
echo:Failed during accel.
Step: 1 from 715
Timer: 328
step_loops: 1
esteps: 339
acc_step_rate:6076
abs_adv_steps_multiplier8:9908
K:300.00
de_float:0.24
mm_D_float:6.32
nom_speed:55.50
nom_rate:6280
axis_steps_per_mm:385.00
delta_adv_steps:339
current_estep_rate:459
current_adv_steps:120
```

**Comment by thinkyhead** [comment]
@Sebastianv650 Do these numbers seem reasonable to you?
```
echo:Failed during accel.
Step: 1 from 715
Timer: 328
step_loops: 1
esteps: 339
acc_step_rate: 6076
abs_adv_steps_multiplier8: 9908
K: 300.00
de_float: 0.24
mm_D_float: 6.32
nom_speed: 55.50
nom_rate: 6280
axis_steps_per_mm: 385.00
delta_adv_steps: 339
current_estep_rate: 459
current_adv_steps: 120
```

**Comment by thinkyhead** [comment]
@Sebastianv650 I wonder if it would be possible to make the updated interrupts-interrupting-interrupts scheme into a conditional behavior, and then test with that behavior disabled. The optimization makes sure bytes aren't lost from the serial UART, but perhaps it contributes to the instability.

Another thing testers can try is to drop the BAUD rate and see if printing is more or less reliable. Perhaps too much time is being tied up in the serial communication and this is exacerbating the behavior — along with `LIN_ADVANCE` incurring extra stepper interrupts.

**Comment by Sebastianv650** [comment]
There are in fact things in you gcode that looks strange. What's the extrusion width you are using for your print? Should it be fixed or are you using different widths for infill, perimeters and so on? As far as I know Cura is still not supporting variable gap fill, therefore even if you would use different extriusion widths for external / internal perimeters and so on I would expect to see distinct values and not a wild range of floating numbers.

While LIN_ADVANCE is designed to handle different extrusion widths, they have to be reasonable. If for example due to an slicer error the extrusion width for a single segment is complete different from the last and next one, this will result in extreme values which would fit to the numbers we see here.

That's what your gcode is looking like, displaying the extrusion width along each segment width a fixed filament diameter of 1.75mm and a layer height of 0.24mm, both taken from your file:
![extrusion widths](https://cloud.githubusercontent.com/assets/10776877/23338293/f6f97fa4-fc06-11e6-8a6e-ff96fc44daa5.JPG)

We can clearly see a favourite extrusion width slightly below 0.4mm. It might be slightly different from the value you set in Cura because a slicer is not calculating with simple rectangle extrusion paths, but I do for this one to keep it simple.
But around that main value, we see spikes of any number between slightly above 0 and around 0.55mm.

So I see the following options to continue:
- Reslice the model with another slicer with your printer settings and try again. But to exclude other error posibilities, also think about the step rate when increasing the FR: Marlin has a limit of 40.000 steps/s, due to the limit to quad stepping I would avoid to go above 20.000 steps/s for printing moves. So for a first step, you might set the print speed to <125mm/s (=20.000/160). I would also do this analysis again to the new gcode file to see what it looks then.
- If you prefer to stick to your Cura version, we could set the ratio used for LIN_ADVANCE to a fixed value according to your real extrusion width and height for test purposis. But I'm not shure if this will work without other errors when the gcode is realy faulty.

We can also test @thinkyhead proposal, disabling the nested interrupts. Most likely you will see other errors then like serial communication errors but as long as you keep your baud rate low it should be OK at least for testing.

From my side, I would prefer to test another slicer first, combined with a second look to the gcode.

**Comment by psavva** [comment]
I printed only of an SD Card...  So, didn't even try printing via USB...
I'm not using anything special on Cura, and don't see how I could have specified different extrusion widths...

**Comment by Sebastianv650** [comment]
It depends on the version of Cura you are using. I'm not using it anymore, but the version I still have installed 18.03 (Lulzbot Version) doesn't have a control option. Therefore the values are all the same as expected:
![cura 18_03](https://cloud.githubusercontent.com/assets/10776877/23341397/0e8f5d42-fc47-11e6-8ce0-7b7994e058b7.JPG)

Here a chart from the current master branch from Prusa Slic3r. Note that the 0.55mm peak is a result of Slic3rs way of doing bringing over sparse infill before the top solid layers, so its alright:
![slic3r](https://cloud.githubusercontent.com/assets/10776877/23341402/22fcea42-fc47-11e6-800b-4e5dcf9a07a7.JPG)

But as far as I know there is a new Cura version where the options are located on the right side of the window, while the old ones had it on the left. I had a short look into a beta version of this a few month ago, and as far as I remeber it had a lot of options compared to the old one. I think there were also options for various line widths, but I can be wrong.

In any case, I guess the version you are using is bugy. I can have a quick look at it if you provide us the Cura version and profile you are using.

**Comment by psavva** [comment]
I'm using the latest Cura Version 2.3.1
Profile is attached:
https://drive.google.com/open?id=0B6kS8tqbO6jVd3ZyMWdNR1pPQm8

However @Sebastianv650 , I think that this is not the real issue here.
The issue is that LIN_ADVANCE is causing the printer to freeze, whereas, without it, it prints happily.

At the end of the day, gcode is a command, and the firmware is supposed to carry out whatever command is sent to it.

I therefore think that the issue which causes the high steps, and freeze is what really should be looked at.

Even if I move to Slicer, It doesn't resolve the issue that this gcode from cura triggers a bug. This is what should really be resolved.

**Comment by james94jeans2** [comment]
2.4 is the latest version of cura

**Comment by psavva** [comment]
I'll update, thanks for letting me know :)

**Comment by Sebastianv650** [comment]
If you execute a bugy *.exe on your PC it can also crash the hole system as a worst case. How should LIN_ADVANCE handle such cases? It can't do what it is told to do, and it wouldn't make any sense even if it could because it was told to do rubish..
Without LIN_ADVANCE it's not that important to have correct extrusion length numbers. If the extruder should do 3 or 5 steps will not harm Marlin, but it might be visible in the print in worst cases or maybe we have a skipped step but nobody notice it.

The "nice" thing: I can reproduce the gcode anomalies with Cura 2.3.1 and your profile. By the way, you can set the line widths under the Quality Tab and they are all set to 0.4mm. According to a short Google search Cura still can't do variable gap fill and such things, so the line widths in the config should be 100% what you get in the gcode.
With Curas layer view I can also not find any lines that look much thinner or thicker than the rest of the model.

Slic3r gcode looks good.
Cura 18.03 (based on 15.02 original Cura) is much better than 2.3.1, but also bad.
I wasn't testing 2.4, but the changelog says nothing about a fixed bug regarding to that.

I also found one possible error source from Cura: It's not thinking about line lengths. The X and Y values of the gcode has 3 places behind the decimal point, sometimes the XY distance of a move is only 0.001mm. On E, we have 5 digits behind the point. But that's not enough to get a precise enough scaling for the extrusion distance if we have a 0.001mm move. Therfore Cura goes to the next possible number, at thats +0.00001mm on E, but that's not what we need in reality.
But thats only one special case, I have no idea what's going wrong on the other places.

I can only think about a work around at the moment: Specify a parameter for LIN_ADVANCE that sets the extrusion width to a fixed number. This parameter could be set within the start-gcode. As only Slic3r uses variable extrusion widths and it also produces nice gcode, it would be OK to have the automatic thing only on Slic3r gcode and fixed numbers for Cura..
Any further idea is welcome.

**Comment by psavva** [comment]
Thank you for your explanation.
I'll vote for setting it in the Firmware as a default, and overwrite it with some GCode in the slice as needed.

**Comment by Sebastianv650** [comment]
OK so I will do the following as long as there are no further ideas:

Configuration_adv.h, LIN_ADVANCE section:
Specify a LIN_ADVANCE_E_D_RATIO = 0 with descrition what it is used for and the formula for calculation (extrusion width * layer height / ((filament diameter / 2)^2 * PI). This will be used as a default value one very printer cold start. 0 means calculate ratio based on gcode values as it is done at the moment in RCBugFix.

Expand the already used M905 command to have additional options beyond K (set K factor) which would be:
D for the filament diameter (D1.75)
W for the extrusion width (W0.4)
H for layer height (H0.2)
If all this three values are passed at once, Marlin will calculate the new E_D_Ratio and use that until a new one is set or the printer is power cycled. This way, you can use a parametric start-gcode so that Cura or any other Slicer should be able to fill in the current values of this 3 parameters when executing or saving the gcode.
If one or more parameters are passed as 0, or if not all 3 values are sent at once, the ratio will be set to 0 (automatic mode as in current RCBugFix).
M905 will return the current K factor and E_D_ratio over serial (if connected).

**Comment by psavva** [comment]
@Sebastianv650 How did you generate the graph for the Extrusion Widths?
Any software, or Excel? 

I will check the same against Slicer, and CURA 2.4

**Comment by Sebastianv650** [comment]
Just Excel and some math. Basicaly I'm importing the gcode to Excel, delete everything that's not G0 or G1 and then create collums for X, Y, E, dE, dXY, dE/dXY and width.

The language will be wrong for you, but maybe you get an idea how to do it:
X=WENNFEHLER(TEIL(A3;FINDEN("X";A3)+1;FINDEN(" ";A3;FINDEN("X";A3))-FINDEN("X";A3)-1);"")
Y=WENNFEHLER(TEIL(A3;FINDEN("Y";A3)+1;WENNFEHLER(FINDEN(" ";A3;FINDEN("Y";A3));99)-FINDEN("Y";A3)-1);"")
E=WENNFEHLER(TEIL(A3;FINDEN("E";A3)+1;9);D2)
dE=D3-D2
dXY=WURZEL((B3-B2)^2+(C3-C2)^2)
dE/dXY=E3/F3
width=(1,75/2)^2*PI()*E3/0,2/F3

**Comment by Sebastianv650** [comment]
@psavva Here is the new branch, including the new option as described above:
https://github.com/Sebastianv650/Marlin/tree/Option-static_E_D_ratio

It's brand new and not tested up to now with real non-test prints. When you test it, remeber to keep the print speed below 125mm/s. If it works up to this speed, you can go higher and see when it crashes ;)

**Comment by psavva** [comment]
@Sebastianv650, 
I am a little unclear how you're expecting the configuration to be defined in the Advanced config file.
Are you Expecting only a value of 0 or 1?
What are all the valid values?  I was under the impression that I should/could use the config file to override the settings if not given by the produced GCODE.

#define LIN_ADVANCE_E_D_RATIO 0 // W * H / ((D / 2)^2 *PI)

Example of valid values for above?
Extrusion width = 0.4mm
Extrusion Height = 0.24mm
D = 1.75mm

From reading the prior comments, I understand it's just a flag to enable/disable the Ration?
ie Valid Values is 0 or 1.

I will try to test using the GCODE example, and specify the ration = 1, which I assume enable :)

**Comment by Sebastianv650** [comment]
No, 1 will lead to realy wired moves. I see I have to redo the info text.
What is expected to go there is the calculated default ratio you get using the formula in the comment.
In your case `0.4*0.24/((1,75/2)^2*PI) = 0.039912.`

So set LIN_ADVANCE_E_D_RATIO to 0.039912.

**Comment by psavva** [comment]
For anybody that would be interested, you can easily use variables to have Cura set start GCODE for you.
This is what I have used:
M905 K300 W{machine_nozzle_size} H{layer_height} D{material_diameter}

The {keywords} above will be replaced by CURA during slicing

**Comment by Sebastianv650** [comment]
I will add that as an example for cura. But with one change: i don't think nozzle size is the right keyword here. There should be something line extrusion width or perimeter extrusion width.

**Comment by psavva** [comment]
Based on your feedback, I found {line_width} exists in the configs...

M905 K300 W{line_width} H{layer_height} D{material_diameter}

Putting this exactly as is in CURA will result in the values being replaced by the keywords.

**Comment by psavva** [comment]
Here is the Excel to Calculate Extrusion Widths if anyone is interested:
https://drive.google.com/open?id=0B6kS8tqbO6jVcUF1bmQzTTVTRzg

**Comment by Sebastianv650** [comment]
@psavva I reworked the description in Configuration_adv.h a little bit, can you have a look at it and decide if it's understandable now? When it's OK and the PR is created, I will add further information like the Cura start code line in the Marlin documentation.

Have you already tried printing with it?

**Comment by psavva** [comment]
I will be out of the country for a week, and will be able to test when I get back...

I've done some preliminary research around CURA extrusion widths and it turns out that it's how CURA has done it since the start.  It's designed that way intentionally. 

Hopefully your enhanced LIN_ADVANCE will play nicely with Cura...

**Comment by Sebastianv650** [comment]
If you find a reason why they do it this way I would love to read that.
There are some points where I can't understand why Cura is behaving as it is, but this one is hard on the edge to call it a bug!

**Comment by hexane360** [comment]
This thread explains some of Cura's extrusion calculations: https://ultimaker.com/en/community/10422-how-is-extrusion-width-calculated

I'm not sure we can assume a constant line width throughout a print. Despite not doing gapfill, Cura still has a few methods to deal with perimeters on thin walls.

**Comment by Sebastianv650** [comment]
@psavva if you already downloaded my branch, please do it again. I just fixed a bug, setting the fixed ratio worked the wrong way around, it enabled the auto detection of the line width instead of disabling it..

**Comment by Sebastianv650** [comment]
@hexane360 as far as I know up to now Cura has no gap fill function. Even if it would, that would be OK if it would generate proper extrusion distances in the gcode. psavva published his Excel file, the graph looks the same as mine so I can confirm he did it right - have a look at it or the picture in [my post](https://github.com/MarlinFirmware/Marlin/issues/5699#issuecomment-282542269). This spikes are not due to variable line widths, I fixed them all to 0.4 during my tests.

**Comment by hexane360** [comment]
I don't think the noise is caused by a variable extrusion rate. I think that's most likely rounding. 

However, what I'm saying is that Cura *does* have some variable extrusion width features, though you are correct it doesn't have gap fill. The thread I linked has a more in depth explanation, but basically if the shell thickness is not an even multiple of the extrusion width, Cura will fudge each line to have a width of (shell thickness)/(num lines), where num lines is something like floor(line width*1.5/shell thickness). Output from cura can have anywhere between 100% to 150% of {line_width}.

I am confident you will see this in practice, although possibly not with the calibration model used here.

**Comment by Sebastianv650** [comment]
> Output from cura can have anywhere between 100% to 150% of {line_width}.

I know that, but with a nozzle of 0.4mm, 0.05mm is far away from 100%-150%.
The gcode tests were done with the dragon model used by psavva, only the first tests I did with the graphs are done with a 20x20x10 cube.

**Comment by hexane360** [comment]
Again, I've never claimed that that's what's causing the inconsistency in extrusion width.

However, it is enough variation to prevent us overriding the extrusion width with one fixed value. On one section of the part, you need much less extrusion than on a different section of the same part. There's no way for one fixed override to emulate this behavior.

If you set the extrusion width to 100% of {line_width}, you'll get underextrusion in sections with 150% of line width.

I think the only solution here is to fix Cura's output.

**Comment by Sebastianv650** [comment]
@hexane360 now I got you :-)
Yes that's true, that's why I put a line in the comment of Configuration_adv.h, stating that this will lead to a wrong pressure adaption if variable line widths or layer heights are used in the print. But I think as long as the gcode output of Cura is what it is, it should be a working solution as long as the extrusion width is not varying too much.

I will add a some words in the Marlin documentation, recommending to fix the line widths in cura all to one value. Luckily that's possible in the recent releases.

**Comment by psavva** [comment]
@Sebastianv650, @thinkyhead 

I have gone ahead and re-tested it.

No Luck... :(

I have inputted the factor directly in the firmware, and did not use the GCODE.

The first problem I noticed is that the very short moves did not extrude properly.
It was like heavy under extrusion for moves < half a cm.
For some moves which were larger than 1 or 2 cm, seems ok, such as the skirt.

After cranking up the FR to 200%, the Printer froze just like it did before...

Serial Monitor Output is here: 
`
start
echo: External Reset
Marlin 1.1.0-RCBugFix

echo: Last Updated: 2016-12-06 12:00 | Author: (none, default config)
Compiled: Mar  5 2017
echo: Free Memory: 3140  PlannerBufferBytes: 1312
echo:V29 stored settings retrieve Y200.00 Zes)
echo:Steps per unit:
echo:  M92 X160.00 Y160.00 Z800.00 E385.00
echo:Maximum feedrates (mm/s):
echo:  M203 X200.00 Y200.00 Z15.00 E100.00
echo:Maximum Acceleration (mm/s2):
echo:  M201 X1500 Y1500 Z500 E5000
echo:Accelerations: P=printing, R=retract and T=travel
echo:  Mime (ms), X=max000.00 T1500.00
echo:Advanced variables: S=Min fe=maximu(mm/s), T=Min travel feedrate (mm/s), B=minimum segment time (ms), X=maximum XY jerk (mm/s),  Z=maximum Z jerk (mm/s),  E=maximum E jerk (mm/s)
echo:  M205 S0.00 T0.00 B20000 X9.00 Y9.00 Z0.40 E7.00
echo:Home offset (mm)
echS1 H240 B110 F0
e.00 Z0.00
Auto Bed Leveling:
echo:  M420 S0
ech.75
terial heatup parameters:
echo:  M145 S0 H190 B50 F0
  M145 S1 H240 B110 F0
echo:PID settings:
echo:  M301 P51.83 I5.02 D133.75
echo:Filament settings: Disabled
echo:  M200 D1.75
echo:  M200 D0
echo:Z-Probe Offset (mm):
echo:  M851 Z-0.60
echo:SD card ok
echo:SD card ok
echo:enqueueing "M23 wb_dra~2.gco"
echo:enqueueing "M24"
echo:Now fresh file: wb_dra~2.gco
File opened: wb_dra~2.gco Size: 4023586
File selected
.
.
.
 T:189.8 /190.0 @:64 W:3
 T:190.0 /190.0 @:61 W:2
 T:190.5 /190.0 @:45 W:1
 T:190.5 /190.0 @:48 W:0
echo:busy: processing
.
.
.
echo:busy: processing
X:160.00 Y:105.00 Z:10.60 E:0.00 Count X:25600 Y:16800 Z:8480
G29 Auto Bed Leveling
echo:busy: processing
.
.
.
echo:busy: processing
Bilinear Leveling Grid:
     0     1     2
 0 +0.41 +0.57 +0.63
 1 +0.17 -0.31 -0.79
 2 +0.49 -0.60 -1.70

X:340.00 Y:290.00 Z:11.86 E:0.00 Count X:54400 Y:46400 Z:8480
X:340.00 Y:290.00 Z:15.00 E:0.00 Count X:54400 Y:46400 Z:8486
X:340.00 Y:290.00 Z:15.00 E:0.00 Count X:54400 Y:46400 Z:8529
echo:busy: processing
echo:busy: processing
`

**Comment by Sebastianv650** [comment]
Can you copy&paste the line you entered in configuration_adv.h together with the values you used for calculation?
It sounds like an inproper value, I just give it another test with a curvy model using a fixed ratio and even with my 20x magnifier there is no problem with the extrusion.
So it might be a typo, mixup of units or we have another problem here.

**Comment by psavva** [comment]
#define LIN_ADVANCE

#if ENABLED(LIN_ADVANCE)
  #define LIN_ADVANCE_K 300

  #define LIN_ADVANCE_E_D_RATIO 0.039912 // The calculated ratio (or 0) according to the formula W * H / ((D / 2) ^ 2 * PI)
                                  // Example: 0.4 * 0.2 / ((1.75 / 2) ^ 2 * PI) = 0.033260135
#endif


I had used the same Values as before, on the same model:
ie: Layer Height 0.24
Filament = 1.75
Nozzle  = 0.4mm

GIT Revision Used: (Today 05 March @ 13:00 20addc6e32e2702f8c70293bb37cf8df5d140968 )

Note: Just removed comments to unclog the post

**Comment by Sebastianv650** [comment]
The ratio is OK. Whatever is causing your under extrusion in detail, this is usualy a symptom of a completely wrong pressure correction. If it's worse than without LIN_ADVANCE enabled at all, the corrections (K and / or the ratio) is much too high. If it's a bit better than without LIN_ADVANCE enabled at all, the corrections is too low.

I guess you did all your K calibration with Cura, meaning K could be off now with the "fixed" workaround.
Can you redo the [flat calibration cube](http://marlinfw.org/docs/features/lin_advance.html#calibrating), ensuring coasting and simmilar things are disabled. Then put them on a scanner or if not available take top looking down photography with a as good resolution as possible and attach them here.
K=100, 200 and 300 should be a good start set.

**Comment by psavva** [comment]
I'll give it a go from start to end.

**Comment by psavva** [comment]
I confirm the issue is resolved.
Thank you @Sebastianv650

**Comment by psavva** [comment]
@thinkyhead, Please close this issue, as it seems I cannot

**Comment by Sebastianv650** [comment]
Thanks for your patience @psavva,this issue was a long one!
Do you got another k value now?

**Comment by psavva** [comment]
Yes, much lower... I'm on 75 now, but am calibrating it further.

**Comment by Sebastianv650** [comment]
When you found your new k value, it would be good to know if the fixed ratio option is still needed for crash free operation :-)

**Comment by LunNova** [comment]
Had what seems to be the same issue, even with a fixed E/D ratio set. Freezes, and busy:processing sent forever.

Tried debugging, stepper::isr stops getting called at this point. That would explain why it's frozen, but I don't know why it's not being called.

Not made much more progress than that as getting it to freeze involved doing a dry run of a print and waiting 15 minutes.

**Comment by LunNova** [comment]
Patch which fixes the freezing but increases the code size by quite a bit. Can probably be made better by someone more familiar with marlin. https://gist.github.com/nallar/d08bd86ec4210ea36d1e4ab824e0e4e5

Problem was that ADV_RATE could return extreme values. 0 would mean advance_isr would be ran every time, and isr would never get called. ADV_NEVER when it shouldn't be would cause advance_isr to not run when it should. Both of these would cause it to stop moving.

**Comment by Sebastianv650** [comment]
I'm curious, what's your microstepping and step/mm settings?

I don't think ADV_NEVER is a real world problem, but you are right about the 0 result. In fact I spent quite some time thinking about a good solution, without finding one as can be seen in the code.

The problem is, as soon as we get a 0 or another nearly 0 value we are over-utilizing the system. Basicaly we can't do LIN_ADVANCE at this point. Real solutions might be to reduce acceleration or jerk, maybe lower the microstepping. Or find ways to live with lower K values.
When you set it to 1, in fact that means setting it to 16 according to the line `NOLESS(OCR1A, TCNT1 + 16);`. In reality, this acts as a limiter that prevents LIN_ADVANCE to execute all the calculated needed steps as it was told due to the given acceleration and K value. While it might work (and protect the printer from freezing) I don't like this because the user has no feedback that the system isn't longer working as it's supposed to. Even if the user want's a higher pressure advance and therefore increase K further and further, he will not see his input in the prints. This will lead to confusion.

Stopping the print or shutting down LIN_ADVANCE by setting K=0 might ruin the non-finished print and is therefore also no good solution.

The only real solution I'm aware of is something I can't do with my knowledge at the moment. It would require a rewrite of the planner. Instead of determining a leading axis (axis with most steps to do) and forcing all other axis to follow, we would need to check if the extruder is able to follow with pressure adoption in mind. This would mean to estimate the maximum extruder speed that will occure due to the pressure correction and if it's too high, reduce acceleration until the maximum speed is inside the limits.
This would be the "easy version", as it keeps the constant acceleration Marlin is using. But it might limit the acceleration more than necessary because the maximum extruder speed might be exceeded only during a short time. Therefore, a variable acceleration or at least an acceleration in in for example 3 steps would be better.
In the end, in both cases it's very likely that the needed planner calculations wouldn't be fast enough on our ATMegas..

In the end, if you can test you changes for some time and you are not seeing other problems, it might be the best option we have at the moment. If other people have this issue, please test it too.
I can't help here as I never get a 0 on my system.

**Comment by LunNova** [comment]
That makes sense. I'm using quite a high steps/mm for the extruder of 912. I should drop it to 16x microstepping.

This issue only occurred occasionally on one part of a print where my slicer made very tight zigzags with lots of small segments. It didn't happen with most prints.

I think for now changing it to return 1 instead of 0 is the least bad option as it prevents a confusing freeze and a ruined print.

**Comment by Sebastianv650** [comment]
If it's working, you could create a PR for this. There are only some small things that should be changed:
- ADV_RATE should become ADV_Rate because the all capital version is a sign if #define.
- This is something to be answered by the more code style pros, but I guess it should be `Stepper::ADV_Rate` because this belongs to the stepper class. Then, it would be also possible to write it as `Stepper::ADV_Rate(uint16_t t, uint16_t l)` as you have direct access to e_steps[], but I guess it's more elegant to have it the way you did.

The code size is only +84 bytes, I think that's no problem!

**Comment by Bob-the-Kuhn** [comment]
Don't be afraid to drop the microstepping.  It doesn't buy you anything as to positional accuracy but it does quiet down the system.
